common/src/test/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImplTest.kt

```diff
@@ -0,0 +1,690 @@
+        setupServer()
+        prepareAuthenticationAnswer()
+        assertTrue(webSocketCore.connect())
+
+        // Message sent by subscribeTo to request for events
+        mockResultSuccessForId(2)
+
+        checkNotNull(
+            webSocketCore.subscribeTo<StateChangedEvent>(
+                SUBSCRIBE_TYPE_SUBSCRIBE_EVENTS,
+                mapOf("event_type" to "state_changed"),
+            ),
+        ).test {
+            webSocketListener.onMessage(mockConnection, """{"id":2, "type":"event", "event":{"event_type":"state_changed", "time_fired":"2016-11-26T01:37:24.265429+00:00", "data": {"entity_id":"light.bed_light"}}}""")
+            assertEquals("light.bed_light", awaitItem().entityId)
+            webSocketListener.onMessage(mockConnection, """{"id":2, "type":"event", "event":{"event_type":"state_changed", "time_fired":"2016-11-26T01:37:24.265429+00:00", "data": {"entity_id":"light.bathroom"}}}""")
+            assertEquals("light.bathroom", awaitItem().entityId)
+        }
+    }
+
+    @Test
+    fun `Given a connection with an existing subscription When subscribeTo is invoked for the same subscription Then it doesn't send a message again and unsubscribes only when the last subscriber unsubscribe`() = runTest {
+        // We need a dedicated scope with unconfined dispatcher that we can control to properly close the shared flow
+        val subscriptionScope = TestScope(UnconfinedTestDispatcher())
+        setupServer(backgroundScope = subscriptionScope)
+        prepareAuthenticationAnswer()
+        assertTrue(webSocketCore.connect())
+
+        // Message sent by subscribeTo to request for events
+        mockResultSuccessForId(2)
+
+        turbineScope {
+            suspend fun createSubscription(name: String) = checkNotNull(
+                webSocketCore.subscribeTo<StateChangedEvent>(
+                    SUBSCRIBE_TYPE_SUBSCRIBE_EVENTS,
+                    mapOf("event_type" to "state_changed"),
+                ),
+            ).testIn(this, name = name).apply {
+                // We should only have one active message even with multiple subscription
+                assertEquals(1, webSocketCore.activeMessages.size)
+                coVerify(exactly = 1) { mockConnection.send(match<String> { it.contains("state_changed") }) }
+            }
+
+            val mainFlow = createSubscription("Main flow")
+
+            webSocketListener.onMessage(mockConnection, """{"id":2, "type":"event", "event":{"event_type":"state_changed", "time_fired":"2016-11-26T01:37:24.265429+00:00", "data": {"entity_id":"light.bed_light"}}}""")
+            assertEquals("light.bed_light", mainFlow.awaitItem().entityId)
+
+            val secondaryFlow = createSubscription("Secondary flow")
+
+            // event are not replayed when subscribing so we should not get any events here
+            secondaryFlow.expectNoEvents()
+
+            // event are received on both flow
```

Comment:
To make sure you get the new message, wouldn't it make more sense to emit a message with a slightly later timestamp here, or a different entity, and check for that?

Reply:
If you mean the timestamp in the message I changed it. But if you mean like adding a delay. Delays in test are skipped to have a better controls. Like everything that is based on time on unit test doesn't rely on real time. `delay(30.s)` will instantly trigger.

Reply:
I was referring to the message contents 👍